Terminal.Gui/Views/FileDialog.cs

```diff
@@ -1,1053 +1,1634 @@
+		}
+		private bool IsCompatibleWithOpenMode (string s, out string reason)
+		{
+			reason = null;
+			if (string.IsNullOrWhiteSpace (s)) {
+				return false;
+			}
+
+			if (!this.IsCompatibleWithAllowedExtensions (s)) {
+				reason = Style.WrongFileTypeFeedback;
+				return false;
+			}
+
+			switch (this.OpenMode) {
 			case OpenMode.Directory:
-				CanChooseFiles = false;
-				CanChooseDirectories = true;
-				break;
+				if (MustExist && !Directory.Exists (s)) {
+					reason = Style.DirectoryMustExistFeedback;
+					return false;
+				}
+
+				if (File.Exists (s)) {
+					reason = Style.FileAlreadyExistsFeedback;
+					return false;
+				}
+				return true;
+			case OpenMode.File:
+
+				if (MustExist && !File.Exists (s)) {
+					reason = Style.FileMustExistFeedback;
+					return false;
+				}
+				if (Directory.Exists (s)) {
+					reason = Style.DirectoryAlreadyExistsFeedback;
+					return false;
+				}
+				return true;
 			case OpenMode.Mixed:
-				CanChooseFiles = true;
-				CanChooseDirectories = true;
-				AllowsMultipleSelection = true;
-				break;
+				if (MustExist && !File.Exists (s) && !Directory.Exists (s)) {
+					reason = Style.FileOrDirectoryMustExistFeedback;
+					return false;
+				}
+				return true;
+			default: throw new ArgumentOutOfRangeException (nameof (this.OpenMode));
 			}
 		}
 
 		/// <summary>
-		/// Gets or sets a value indicating whether this <see cref="Terminal.Gui.OpenDialog"/> can choose files.
+		/// Changes the dialog such that <paramref name="d"/> is being explored.
 		/// </summary>
-		/// <value><c>true</c> if can choose files; otherwise, <c>false</c>.  Defaults to <c>true</c></value>
-		public bool CanChooseFiles {
-			get => dirListView.canChooseFiles;
-			set {
-				dirListView.canChooseFiles = value;
-				dirListView.Reload ();
+		/// <param name="d"></param>
+		/// <param name="addCurrentStateToHistory"></param>
+		/// <param name="setPathText"></param>
+		/// <param name="clearForward"></param>
+		internal void PushState (IDirectoryInfo d, bool addCurrentStateToHistory, bool setPathText = true, bool clearForward = true)
+		{
+			// no change of state
+			if (d == this.State?.Directory) {
+				return;
+			}
+			if (d.FullName == this.State?.Directory.FullName) {
+				return;
+			}
+
+			PushState (new FileDialogState (d, this), addCurrentStateToHistory, setPathText, clearForward);
+		}
+
+		private void RefreshState ()
+		{
+			State.RefreshChildren ();
+			PushState (State, false, false, false);
+		}
+
+		private void PushState (FileDialogState newState, bool addCurrentStateToHistory, bool setPathText = true, bool clearForward = true)
+		{
+			if (State is SearchState search) {
+				search.Cancel ();
+			}
+
+			try {
+				this.pushingState = true;
+
+				// push the old state to history
+				if (addCurrentStateToHistory) {
+					this.history.Push (this.State, clearForward);
+				}
+
+				this.tbPath.Autocomplete.ClearSuggestions ();
```

Comment:
Does stuff like this really belong in `Redraw`? It's a yellow-flag to me when I see `Redraw` implementations changing state. Is there another event (`LayoutComplete`?) where this stuff can be done?

I now believe a litmus test for good `View` design (and a test of how good our public API is) is for `View` overrides to not have to override `Redraw`. Took me a long time to get to this point ;-)

Reply:
This code is in `PushState` method not `Redraw`.  Redraw currently only has the Feedback drawing and that will be moved to Margin 🚀